modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1273,7 +1275,6 @@ def binary_operation(cls, axis, left, func, right):
         return result if axis else result.T
 
     @classmethod
-    @wait_computations_if_benchmark_mode
```

Comment:
Why is this removed?

Reply:
Don't the lines added above resolve the issue?

Reply:
both changes should be made:
1. We shouldn't wrap None-returning methods with this decorator, as they do not return any waitable objects
2. If we did an error and wrapped None-returning method, we still shouldn't fail but just do a return without waiting anything

Reply:
What if we want to measure the following case when BenchmarkMode is enabled.

```python
t0 = time()
pickle.dumps(df) # __reduce__ -> finalize
t1 = time()
print(t1 - t0)
```
Would that give correct timing?

Reply:
This will no longer be correct, since drain call queue isn't blocking. You'd have to add something that waits on each partition, either by modifying `finalize` to check if benchmark mode is on and doing a wait, or by modifying finalize to return the partitions (perhaps always, perhaps only if benchmark mode is on) and waiting on them.

Reply:
Fixed this, now the decorator also seeks for partitions to await inside the function's arguments if the function returns None

Reply:
> This will no longer be correct, since drain call queue isn't blocking. You'd have to add something that waits on each partition, either by modifying `finalize` to check if benchmark mode is on and doing a wait, or by modifying finalize to return the partitions (perhaps always, perhaps only if benchmark mode is on) and waiting on them.

It turns out that there is a mismatch with what is written in the docstring of `finalize` method. So we should add a loop where method `wait` will be explicitly called on each partition.

In this case, it will be sufficient to simply remove the decorator (without making changes to it).

Reply:
What do you mean exactly? The docstring says "Perform all deferred calls on partitions.", but not "Perform all deferred calls on partitions and wait until all computations are complete."

Reply:
The current implementation should be understood as: "Trigger all deferred computations on partitions". 

To write "perform" in the docstring, the function should wait for the results. Isn't that what the function is for?

Reply:
Nope, the method is meant to perform all deferred calls, which have been added to the call queues, on partitions in asynchronous way.

Reply:
if so, the simplest way here is return 2d ndarray with partitions

Reply:
Even though this is the simplest way to fix the issue, I think this is the right way and the partition manager should always return updated partitions to construct a new core Modin dataframe to avoid any further side effects.

Reply:
cc @dchigarev

Reply:
done, now `PartitionManager.finalize` method returns passed partitions after draining call queues on them

Reply:
What I mean is we should also construct a new core Modin dataframe, as well as update inplace the high-layer DataFrame.

Also, I wonder whether we need the new logic for the decorator as it will hide similar errors to one be resolved in this PR? Because we assume the partition manager should always return updated partitions.

Reply:
> What I mean is we should also construct a new core Modin dataframe, as well as update inplace the high-layer DataFrame.

Not sure about that. What `finalize` is intended to do is to ["Finalize <b><i>CONSTRUCTING</b></i> the dataframe calling all deferred functions which were used to build it"](https://github.com/modin-project/modin/blob/270fd697a82d051f0ab2c54b70b74296699bdcf2/modin/core/storage_formats/base/query_compiler.py#L275). The `finalize` method when called against a finalizing object, have to mutate the actual object, without creating a new one that is considered to be finalized.

> I wonder whether we need the new logic for the decorator as it will hide similar errors to one be resolved in this PR? Because we assume the partition manager should always return updated partitions.

It won't hide similar errors, it will just expand a set of use-cases when the decorator could be applied (now it's possible to use it for inplace methods as well). For the similar-error cases there's now a check, that raises a proper error if so happened. [[1]](https://github.com/dchigarev/modin/blob/2eea67dc78353f5fccfd93d7ba498d83eca9173c/modin/core/dataframe/pandas/partitioning/partition_manager.py#L84-L87)

Reply:
> BTW, notice that even `PartitionManager.finalize` does not create new finalized partitions, but just returns exactly the same numpy array that was passed as `partitions`.

Please note, the content of returned partitions may become different so I was wondering a case when metadata might be unsynced. However, because we take care enough of syncing metadata at the core Dataframe layer, there shouldn't be such a case and the flow will be correct until partitions (NumPy array) object is immutable. So I think we should probably be explicit and assign new returned partitions to `_partition` field of the core Dataframe. Does this make sense?

Reply:
> So I think we should probably be explicit and assign new returned partitions to _partition field of the core Dataframe. Does this make sense?

Don't know. I'm a bit tensed about this approach as well as about returning "finalized" partitions from `PartitionManager.finalize`. IMHO, assigning new `._partitions` is redundant and would confuse developers as the line literally means `self._partitions = self._partitions` but creates an impression of newly created "finalized" array of partitions. I don't want to be a blocker here as the discussion we had in the PR is already quite big, but I still not really like the proposed approach because of its counterintuitivity.

So would like to hear other opinions from @modin-project/modin-core about this thread and I will just agree with the majority.

The questions are:
1. [`PartitionManager.finalize`](https://github.com/modin-project/modin/blob/270fd697a82d051f0ab2c54b70b74296699bdcf2/modin/core/dataframe/pandas/partitioning/partition_manager.py#L1275-L1286) modifies the passed partition's state by calling the deferred calls on them (no new partitions produced during this), is this okay to return the passed partitions in the result for the sake of interface aligning with other manager's method? My concern is that it creates a false impression of the function's user that the returned array is a new one.
    ```python
    @classmethod
    def finalize(cls, partitions):
        [partition.drain_call_queue() for partition in partitions.flaten()]
        return partitions # return exactly the same object that was passed into the func
    ```
    (we can fix the actual issue (#4273) without that by just inspecting the wrapped method's argument for partitions to wait)
2. If we decide to return partitions from `PartitionManager.finalize`, should we then update `._partitions` with the new ones (actually exactly the same objects) that were returned from `PartitionManager.finalize`? My concern is that it could be really misleading, as the statement would literaly mean `self._partitions = self._partitions`:
    ```python
    def finalize(self):
        self._partitions = self._partition_mgr_class.finalize(self._partitions)
    ```

Reply:
> the line literally means `self._partitions = self._partitions`

As I said, this will be the same until `partitions` (NumPy array) object is immutable.

Reply:
@YarShev what if `finalize` function does not return anything, as before? Just add a mode check and call `wait` method on all partitions if necessary. I don't see any problems that in some cases the partition manager does not return anything. 

It seems to me that we can create more situations with possible side effects (for example, sharing an array with several dataframes) if we return an array here.

Reply:
> @YarShev what if `finalize` function does not return anything, as before? Just add a mode check and call `wait` method on all partitions if necessary. I don't see any problems that in some cases the partition manager does not return anything.

@anmyachev, imagine a case when the child core Dataframe contains partitions in an immutable object (or we would decide that for the `PandasDataframe`). If we do not return partitions and not assign them to the respective field, the `PandasDataframe` or its child will contain unchanged partitions.

> It seems to me that we can create more situations with possible side effects (for example, sharing an array with several dataframes) if we return an array here.

Not sure I got you. The array will contain updated partitions.

This discussion is really huge. I wish we converged to a conclusion.

cc @modin-project/modin-core

Reply:
I imagined, but in this case we need to change the data structure, as numpy array is mutable. Such a drastic change looks beyond the scope of this pull request.

As long as partitions object is mutable, by returning them we can assign them to another dataframe, which will lead to a situation where changes in one object will affect another.

I wish the same :)

Reply:
> I imagined, but in this case we need to change the data structure, as numpy array is mutable. Such a drastic change looks beyond the scope of this pull request.

Of course, I am saying about the case when the container for partitions is immutable. We do not want to change it now and probably in future for the `PandasDataframe`.

> As long as partitions object is mutable, by returning them we can assign them to another dataframe, which will lead to a situation where changes in one object will affect another.

We would assign partitions to the `self` dataframe, but not to another.

@devin-petersohn, @RehanSD, @vnlitvinov, thoughts?